### ISSUE TITLE
Handle covariant callbacks

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#27487]
Flutter has a few cases where they are trying to use functions in a way that is statically unsound. Almost all of them are around `GestureRecognizer` and look roughly like:

``` dart
typedef GestureRecognizer GestureRecognizerFactory(GestureRecognizer recognizer);

Widget build(BuildContext context) {
  Map<Type, GestureRecognizerFactory> gestures = <Type, GestureRecognizerFactory>{};

  gestures[TapGestureRecognizer] = (TapGestureRecognizer recognizer) {
    ...
  };
}
```

Here, that map set isn't sound because the function literal takes a `TapGestureRecognizer`, not a `GestureRecognizer`. The former is a subtype of the latter, so this isn't statically safe.

Before Flutter can become strong mode clean, we need to address this. We need to decide how/if we want to change the language to handle code like this. And then, once we have, we need to implement and spec it.

**Comment by leafpetersen** [comment]
@sethladd @Hixie @abarth

**Comment by lrhn** [comment]
How does the return type of the factory function type vary?
Could the factory type be declared as:
```dart
typedef T GestureRecognizerFactory<T extends GestureRecognizer>(T recognizer);
```
or some other generic function type? Or is there no structure to how the argument and return type vary?

One solution I see is to just wrap the function in a class.
```dart
class GestureRecognizerFactoryBox<T> {
  final GestureRecognizerFactory<T> _factory;
  GestureRecognizerFactoryBox(this._factory);
  T create(T recognizer) => _factory(recognizer);
}
```
Then the class is correctly typed, and because it's a class type, not a function type, it allows covariant subtyping. That means that you can store a `GestureRecognizerFactoryBox<TapGestureRecognizer>` in the map and call it again without running into problems.
The `create` function has a contravariant occurrence of the type parameter, so we will insert a check, but that's expected for generic classes like that.
  
It sucks somewhat to have to make a class instance to keep a single function, but the existing coide is type-unsafe as written, and we don't allow type-unsafe code. I don't think we should compromise that for code like this.


Alternatively, don't use a plain normal map, use a specialized class with type-aware `lookup` and `add` operations which allow the specific type variations needed for handling what is basically dependently typed code.

**Comment by Hixie** [comment]
> Could the factory type be declared as:
>
> ```dart
> typedef T GestureRecognizerFactory<T extends GestureRecognizer>(T recognizer);
>```

Yes, that would be an accurate description of the constructors. The reason we don't do that is that there's no way to specialise the type per entry in the map, so there's not much gain to be had from it.

> It sucks somewhat to have to make a class instance to keep a single function, but the existing code is type-unsafe as written, and we don't allow type-unsafe code. I don't think we should compromise that for code like this when there are simple workarounds.

The ability to write this kind of code is part of what made Dart 1.0 great. I would be very sad to lose it.

**Comment by munificent** [comment]
> Alternatively, wrap the function when it's added to the map:
> 
> ```dart
> void addGestureRecognizerFactory<T extends GestureRecognizer>(Type<T> type, 
>                                                               T factory(T recognizer)) {
>   _map[type] = (GestureRecognizer r) => factory(r as T);
> }
> ```  

Yup. I fleshed out a proof of concept of that here: https://github.com/munificent/flutter/commit/8def6c700d996efd1aa3c81be620555bf6001529

**Comment by leafpetersen** [comment]
status: under discussion.

**Comment by Hixie** [comment]
I think if we had true generic methods (i.e. generic methods were at runtime we could inspect the type argument) then we could rewrite this code to not require covariant callbacks.

**Comment by floitschG** [comment]
Not planned for 1.50.

**Comment by Hixie** [comment]
@cbracken ran into a case today where he wanted to say `covariant` in a typedef.

**Comment by srawlins** [comment]
cc @matanlurey we've discussed this one a lot 😄

**Comment by eernstg** [comment]
`covariant` in a typedef does not fit well:

With `covariant` on a formal parameter declaration in an instance method, we will get dynamic checks for the actual argument, and we get the permission to add further covariance on the same parameter in all subtypes.

In order to ensure expression soundness we give a tear-off of such a method the reified parameter type `Object` for each parameter which is covariant. In other words, we allow such a tear-off to masquerade as a function that accepts _everything_ for that parameter.

However, we did not allow `covariant` as part of a function type (say, in a `typedef`) at all: Among other things, a function type is used as a type annotation in declarations of variables, parameters, returned values, etc. In all those cases the relevant function object exists already. This means that we cannot retroactively change its dynamic type (short of creating a wrapper, which is an anomaly because the original function and the wrapper would _have_ to have different identity, because they have different types, and assignment doesn't otherwise create a new object). I believe that it's just not a good idea to try to make `covariant` a property of the type of a function.

However, we could certainly allow functions in general (including function literals) to obtain the same kind of flexibility, using the same syntax:

```
void globalFunction(covariant int x) => x.isEven; // Safe, will check that `x is int`.
void Function(dynamic) f = globalFunction as dynamic; // OK, is `void Function(Object)`.

main() {
  f = (covariant int x) => x.isEven; // Also fine, is `void Function(Object)`.
  globalFunction('Not an int'); // Compile-time error.
  ((covariant int x) => x.isEven)('Not an int'); // Compile-time error.
}
```

The underlying principle is that function types (static and dynamic) have no idea that any particular function has a covariant parameter, it's completely an implementation detail for that function: It announces (in its dynamic type) that it will accept arbitrary arguments (`Object`), but then it will check dynamically that it actually gets something more specific (here: `int`), and the body of the function can then use that more specific type. But the rest of the world will not be able to see these things, they just see a function accepting an `Object`, according to the dynamic type.

The static type preserves the declared type of the covariant parameter. This basically prevents "stupid" invocations where we can predict that the dynamic check will fail (like `globalFunction('Not an int')`).

So if you actually want `covariant` to be part of a function type then we'd need to reconsider the entire story about what a covariant parameter is. I'm not saying that this is impossible, just that it's pretty far away from the thinking that we used when this feature was introduced.

**Comment by stan-at-work** [comment]
This would be nice, then we would not have to do this:

```dart
      identifier: CoreAuthTokenRefreshAccessTokenIdentifier(
        withTrigger:
            (CoreAuthTokenManagerState previousState, CoreAuthTokenManagerState currentState) {
                  return currentState.accessToken.isExpired == false;
                }
                as bool Function(CoreBaseState<dynamic>, CoreBaseState<dynamic>),
      ),
```

But just do this:

```dart
      identifier: CoreAuthTokenRefreshAccessTokenIdentifier(
        withTrigger:
            (covariant CoreAuthTokenManagerState previousState, covariant CoreAuthTokenManagerState currentState) {
                  return currentState.accessToken.isExpired == false;
            }
      ),
```

**Comment by eernstg** [comment]
@stan-at-work, thanks for the example! It's somewhat complicated to explain why this feature is appropriate and useful, but let me try. ;-)

### Covariant parameters can be helpful

The ability to use the modifier `covariant` on some formal parameter declarations in a Dart program is often considered to be questionable. It may be seen as similar to using the type `dynamic` because that's easier, thus lowering the level of correctness support in the software.

However, I consider judicious use of `covariant` to be a useful tool in the ongoing effort to support type safety well in complex software. That is, it doesn't degrade the level of static type safety, it _helps developers writing correct code_ even in situations where the static type system doesn't capture the actual structure of the code well enough to be able to recognize it as type safe.

Most examples are complex and wouldn't fit in a comment here, but here's a simple one that has relevant properties: 

```dart
abstract class Vehicle {
  final int id;
  Vehicle(this.id);
  void start(covariant Operator _);
}

class Car extends Vehicle {
  Car(super.id);
  void start(Driver d) => print("Car $id was started by driver ${d.id}!");
}

class Bicycle extends Vehicle {
  Bicycle(super.id);
  void start(Cyclist c) => print("Bicycle $id was started by cyclist ${c.id}!");
}

abstract class Operator {
  final int id;
  Operator(this.id);
}

class Driver extends Operator {
  Driver(super.id);
}

class Cyclist extends Operator {
  Cyclist(super.id);
}

final owner = <Vehicle, Operator>{};

void main() {
  owner[Car(1)] = Driver(1);
  owner[Bicycle(1)] = Cyclist(1);
  owner[Car(2)] = Driver(2);
  
  for (Vehicle vehicle in owner.keys) {
    Operator operator = owner[vehicle]!;
    vehicle.start(operator); // A run-time type check on `operator` occurs here.
  }
}
```

The point is that we're storing objects in the map in a way which is lossy in terms of static typing (we're forgetting that we're dealing with a `Car` and a `Driver`, reverting to `Vehicle` and `Operator`), but we still want to maintain that it is OK for a `Driver` to drive a `Car` and it's OK for a `Cyclist` to ride a `Bicycle`, but it's not necessarily safe for a `Driver` to drive a `Bicycle`, and it may even be illegal for a `Cyclist` to drive a `Car`.

There are lots of ways to create a more elaborate setup such that we can preserve the type information for a specific usage. For example, we could use a `List<Pair>` rather than a `Map`, where `Pair` is a sealed class with subtypes `CarPair` and `BicyclePair` where the types `Car` etc. are specified statically, and then we could remove the method `start` from the superclasses `Vehicle` and `Operator`, and then `start` in `Car` and `Bicycle` wouldn't have to have a `covariant` parameter. We'd then have something like the following `for` statement:

```dart
for (Pair pair in owner) {
  switch (pair) {
    case CarPair():
      Car car = pair.first;
      Driver driver = pair.second;
      car.start(driver); // No run-time type checks needed.
    case BicyclePair():
      Bicycle bicycle = pair.first;
      Cyclist cyclist = pair.second;
      bicycle.start(cyclist); // No run-time type checks needed.
  }
}
```

However, all these tricks are complex, and they don't necessarily scale well (for instance, we'd need to create more `...Pair` classes, and every location where we wish to invoke `start` would have to have a new case for each new `...Pair` class).

This illustrates that we can have regularities in programs that are not captured by a (reasonably simple and scalable) modeling in the type system. In the case where we rely on a run-time type check (requested by the modifier `covariant`, as shown in the example above), I'd claim that it is helpful to know that we're following a special discipline that the type system doesn't know about (in the example above, we mustn't ever do `owner[Car(...)] = Cyclist(...)`).

### Covariant parameters in function literals

Returning to this issue, the ability to have function literals with `covariant` parameters is just another example of this phenomenon: We're relying on the function literal to be invoked on arguments with the specified types, but we're aware of the fact that the function literal is passed to code that expects less specific arguments. The software will be maintained to ensure that the function is actually called with suitable actual arguments, even though the type system doesn't know how we're achieving that end.

In order to help developers recognize that we're working in the context of a special, application specific programming discipline that makes it safe to assume more about the actual arguments than that which is obtained from static type analysis, we can use the modifier `covariant` on the parameters:

```dart
  identifier: CoreAuthTokenRefreshAccessTokenIdentifier(
    withTrigger: (_, covariant CoreAuthTokenManagerState currentState) =>
        !currentState.accessToken.isExpired,
  ),
```

Note that we're only declaring the type (and the property of being `covariant`) on the parameter `currentState`, because we don't use the first parameter at all. So we're going beyond what the type system can prove (which is what `covariant` says), but we're only doing it to the minimal extent where we actually need it.

(Of course, it's possible that we could catch some additional bugs by insisting that `previousState` must also be a `CoreAuthTokenManagerState`, but in that case I'd probably recommend checking `previousState is CoreAuthTokenManagerState` in an assertion. That's a matter of style.)

Without `covariant` we'd have to write something like this:

```dart
  identifier: CoreAuthTokenRefreshAccessTokenIdentifier(
    withTrigger:
        (_, currentState) {
          final state = currentState as CoreAuthTokenManagerState;
          return !state.accessToken.isExpired;
        },
  ),
```

This has essentially the same semantics, but the difference is that the form that uses a `covariant` parameter makes it an API property of this function literal that it is expected to be treated with this special discipline whereby all invocations will actually pass a `CoreAuthTokenManagerState` as the second argument. The type checker doesn't know how that would be achieved, but developers would have written `CoreAuthTokenRefreshAccessTokenIdentifier` in such a way that it is actually true. This special (beyond-the-type-system) requirement is maintained collectively by the function literal and `CoreAuthTokenRefreshAccessTokenIdentifier`, and that's the reason why it is essentially an API property rather than a secret implementation detail in the body of the function literal.

The point I'm making is that this makes it _more conceptually appropriate_ to express this behavior using the modifier `covariant` in the header of the function literal than it is to hide it in the body using an `as` expression.

So let's keep this proposal on the table.
